pkg/kubectl/cmd/util/cmdtools.go

```diff
@@ -0,0 +1,265 @@
+// editor and diff tools
+type CmdTool struct {
+	Name    CmdName
+	Args    []string
+	IsShell bool
+}
+
+// NewDefaultCmdTool creates a struct CmdTool that uses the OS environment to
+// locate the editor or diff program, looking at environment variable to find
+// the proper command line then passed to the user's shell for execution. if
+// no environment variable specified, it will use some default command directly.
+func NewDefaultCmdTool(name CmdName, envs []string) CmdTool {
+	exec, shell := defaultEnv(name, envs)
+	return CmdTool{
+		Name:    name,
+		Args:    exec,
+		IsShell: shell,
+	}
+}
+
+func defaultEnvShell() []string {
+	shell := os.Getenv("SHELL")
+	if len(shell) == 0 {
+		shell = platformize(defaultShell, windowsShell)
+	}
+	flag := "-c"
+	if shell == windowsShell {
+		flag = "/C"
+	}
+	return []string{shell, flag}
+}
+
+func defaultEnv(cmdName CmdName, envs []string) ([]string, bool) {
+	var cmd string
+	for _, env := range envs {
+		if len(env) > 0 {
+			cmd = os.Getenv(env)
+		}
+		if len(cmd) > 0 {
+			break
+		}
+	}
+
+	if len(cmd) > 0 {
+		// rather than parse the shell arguments ourselves, punt to the shell
+		shell := defaultEnvShell()
+		return append(shell, cmd), true
+	}
+
+	switch cmdName {
+	case EditorCmd:
+		cmd = platformize(defaultEditor, windowsEditor)
+	case DiffCmd:
+		cmd = checkGitDiffInstalled(gitDiff, normalDiff)
+	}
+	return []string{cmd}, false
+}
+
+func (t CmdTool) args(path string) []string {
+	// there is a slice operation in this function could change
+	// t.Args permanently, so we need to do a deep copy here first.
+	args := make([]string, len(t.Args))
+	copy(args, t.Args)
+
+	if t.IsShell {
+		last := args[len(args)-1]
+		args[len(args)-1] = fmt.Sprintf("%s %q", last, path)
+	} else {
+		args = append(t.Args, path)
+	}
+	return args
+}
+
+// Launch opens the described or returns an error. The TTY will be protected, and
+// SIGQUIT, SIGTERM, and SIGINT will all be trapped.
+func (t CmdTool) Launch(paths []string) error {
+	cmdPrefix := make([]string, len(t.Args))
+	copy(cmdPrefix, t.Args)
+	if len(t.Args) == 0 {
+		return fmt.Errorf("no %s defined, can't open %s", t.Name, paths[0])
+	}
+	for _, path := range paths {
+		abs, err := filepath.Abs(path)
+		if err != nil {
+			return err
+		}
+		t.Args = t.args(abs)
+	}
+	cmd := exec.Command(t.Args[0], t.Args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	glog.V(5).Infof("Opening file with %s %v", t.Name, t.Args)
+	if err := (term.TTY{In: os.Stdin, TryDev: true}).Safe(cmd.Run); err != nil {
+		if err, ok := err.(*exec.Error); ok {
+			if err.Err == exec.ErrNotFound {
+				return fmt.Errorf("unable to launch the %s %q", t.Name, strings.Join(cmdPrefix, " "))
+			}
+		}
+
+		//diff command will return 1 when compare different
+		if t.Name == "diff" {
+			if _, ok := err.(*exec.ExitError); ok {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("unknown error with the %s %q", t.Name, strings.Join(cmdPrefix, " "))
```

Comment:
Why not do this?

```go
return fmt.Errorf("error running %v %v: %v", t.Name, cmdPrefix, err)
```

In cases where you have access to the error, it's better to include some information from that error in your result.